src/dataflow/source/mod.rs

```diff
@@ -52,6 +57,50 @@ pub struct SourceConfig<'a, G> {
     pub consistency: Consistency,
 }
 
+#[derive(Clone, Serialize, Deserialize)]
+/// A record produced by a source
+pub struct SourceOutput<K, V>
+where
+    K: Data,
+    V: Data,
+{
+    /// The record's key (or some empty/default value for sources without the concept of key)
+    pub key: K,
+    /// The record's value
+    pub value: V,
+    /// The position in the source, if such a concept exists (e.g., Kafka offset, file line number)
+    pub position: Option<i64>,
+}
+
+impl<K, V> SourceOutput<K, V>
+where
+    K: Data,
+    V: Data,
+{
+    /// Build a new SourceOutput
+    pub fn new(key: K, value: V, position: Option<i64>) -> SourceOutput<K, V> {
+        SourceOutput {
+            key,
+            value,
+            position,
+        }
+    }
+}
+impl<K, V> SourceOutput<K, V>
+where
+    K: Data + Hashable<Output = u64> + Serialize + for<'a> Deserialize<'a> + Send + Sync,
+    V: Data + Hashable<Output = u64> + Serialize + for<'a> Deserialize<'a> + Send + Sync,
```

Comment:
Don't worry at all about this at the moment, but you can move the `Hashable` constraint on to each of the methods below, which should allow e.g. settings with a `Hashable` key partitioned by key, even if the value is not.

Reply:
Excellent point